core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -177,6 +190,41 @@ public RecoveryResponse recoverToTarget() throws IOException {
     }
 
     /**
+     * Determines if the source translog is ready for a sequence-number-based peer recovery. The main condition here is that the source
+     * translog contains all operations between the local checkpoint on the target and the current maximum sequence number on the source.
+     *
+     * @param translogView a view of the translog on the source
+     * @return {@code true} if the source is ready for a sequence-number-based recovery
+     * @throws IOException if an I/O exception occurred reading the translog snapshot
+     */
+    boolean isTranslogReadyForSequenceNumberBasedRecovery(final Translog.View translogView) throws IOException {
+        final long startingSeqNo = request.startingSeqNo();
+        final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
+        logger.trace("starting: [{}], ending: [{}}", startingSeqNo, endingSeqNo);
+        if (startingSeqNo - 1 <= endingSeqNo) {
```

Comment:
hmm. The snapshot and restore aspect is interesting. I think we need to deal with it separately as it basically violates all assumptions on the local/global checkpoint semantics. I added a note to https://github.com/elastic/elasticsearch/issues/10708 about it.  Potentially just force flush so no doc level recovery will be possible. Can you add a norelease here with a note about how a restore  can cause havoc?